src/main/java/com/icthh/xm/ms/entity/service/impl/XmEntityServiceImpl.java

```diff
@@ -189,6 +198,90 @@ public XmEntity saveXmEntity(XmEntity xmEntity) {
         return xmEntityRepository.save(xmEntity);
     }
 
+    private void assertEntitySourcesMaxLinkValue(XmEntity xmEntity) {
+        if (xmEntity.getSources().isEmpty()) {
+            return;
+        }
+        Set<Link> newLinks = xmEntity.getSources().stream().filter(Link::isNew).collect(toSet());
+        Set<Link> newUniqLinks = new TreeSet<>((l1, l2) -> {
+            if (l1.getTypeKey().equals(l2.getTypeKey()) && l1.getSource().getId().equals(l2.getSource().getId())) {
+                return 0;
+            }
+            return 1;
+        });
+        newUniqLinks.addAll(newLinks);
+        List<Long> sourceIds = newUniqLinks.stream().map(Link::getSource).map(XmEntity::getId).distinct().collect(toList());
+        List<String> sourceTypeKey = xmEntityRepository.findAllStateProjectionByIdIn(sourceIds)
+            .stream().map(XmEntityStateProjection::getTypeKey).collect(toList());
+        Map<Long, String> sourceIdsAndTypeKeys = IntStream.range(0, (sourceIds.size())).boxed()
+            .collect(toMap(sourceIds::get, sourceTypeKey::get));
+        newUniqLinks.forEach(it -> linkCount(it.getTypeKey(),
```

Comment:
can we replace loop of sql queries by groupBy?
https://stackoverflow.com/questions/6955993/how-to-get-select-values-and-count-value-with-jpa-group-by